<?xml version="1.0" encoding="UTF-8"?>
<root BTCPP_format="4">
  <BehaviorTree ID="Untitled">
    <ReactiveSequence>
      <SubGameStatus topic_name="game_status"
                     game_status="{game_status}"/>
      <SubRobotStatus topic_name="robot_status"
                      robot_status="{robot_status}"/>
      <SubRfidStatus topic_name="rfid_status"
                     rfid_status="{rfid_status}"/>
      <WhileDoElse>
        <IsGameTime message="{game_status}"
                    game_progress="4"/>
        <WhileDoElse>
          <IsStatusBest message="{robot_status}"/>
          <SendGoal goal_pose="5;-3.5;0; 0;0;0;1"
                    action_name="navigate_to_pose"/>
          <WhileDoElse>
            <IsStatusBad message="{robot_status}"/>
            <SendGoal goal_pose="0;0;0; 0;0;0;1"
                      action_name="navigate_to_pose"/>
            <WhileDoElse>
              <AsyncSequence>
                <IsStatusMiddle message="{robot_status}"/>
                <IsRfidNotDetected message="{rfid_status}"/>
              </AsyncSequence>
              <SendGoal goal_pose="5;-3.5;0; 0;0;0;1"
                        action_name="navigate_to_pose"/>
              <Sleep msec="0"/>
            </WhileDoElse>
          </WhileDoElse>
        </WhileDoElse>
        <SendGoal goal_pose="0;0;0; 0;0;0;1"
                  action_name="navigate_to_pose"/>
      </WhileDoElse>
    </ReactiveSequence>
  </BehaviorTree>

  <!-- Description of Node Models (used by Groot) -->
  <TreeNodesModel>
    <Action ID="IsGameTime"
            editable="true">
      <input_port name="message"
                  default="{game_status}"/>
      <input_port name="game_progress"
                  default="4"/>
    </Action>
    <Action ID="IsRfidNotDetected"
            editable="true">
      <input_port name="message"
                  default="{rfid_status}"/>
    </Action>
    <Action ID="IsStatusBad"
            editable="true">
      <input_port name="message"
                  default="{robot_status}"/>
    </Action>
    <Action ID="IsStatusBest"
            editable="true">
      <input_port name="message"
                  default="{robot_status}"/>
    </Action>
    <Action ID="IsStatusMiddle"
            editable="true">
      <input_port name="message"
                  default="{robot_status}"/>
    </Action>
    <Action ID="SendGoal"
            editable="true">
      <input_port name="goal_pose"
                  default="0;0;0; 0;0;0;1"/>
      <input_port name="action_name"
                  default="navigate_to_pose"/>
    </Action>
    <Action ID="SubGameStatus"
            editable="true">
      <input_port name="topic_name"
                  default="game_status"/>
      <output_port name="game_status"
                   default="{game_status}"/>
    </Action>
    <Action ID="SubRfidStatus"
            editable="true">
      <input_port name="topic_name"
                  default="rfid_status"/>
      <output_port name="rfid_status"
                   default="{rfid_status}"/>
    </Action>
    <Action ID="SubRobotStatus"
            editable="true">
      <input_port name="topic_name"
                  default="robot_status"/>
      <output_port name="robot_status"
                   default="{robot_status}"/>
    </Action>
  </TreeNodesModel>

</root>